pkg/chains/signing.go

```diff
@@ -57,22 +78,58 @@ type TaskRunSigner struct {
 	SecretPath        string
 }
 
-// IsSigned determines whether a TaskRun has already been signed.
-func IsSigned(tr *v1beta1.TaskRun) bool {
+// SignedState returns TaskRun state.
+func SignedState(tr *v1beta1.TaskRun) State {
 	signature, ok := tr.ObjectMeta.Annotations[ChainsAnnotation]
 	if !ok {
-		return false
+		return Unknown
+	}
+
+	if value, ok := signedStates[signature]; !ok {
+		return Unknown
+	} else {
+		return value
 	}
-	return signature == "true"
+}
+
+// returns true if:
+// the signed annotation isn't there, the failed annotation isn't there, and retries<3
+func RetryAgain(tr *v1beta1.TaskRun) bool {
+	_, ok := tr.ObjectMeta.Annotations[ChainsRetriesAnnotation]
+	return ok
 }
 
 // MarkSigned marks a TaskRun as signed.
 func MarkSigned(tr *v1beta1.TaskRun, ps versioned.Interface, annotations map[string]string) error {
+	return markannotation(tr, ps, annotations, ChainsAnnotation, Pass)
+}
+
+// MarkFailed marks a TaskRun as signed.
+func MarkFailed(tr *v1beta1.TaskRun, ps versioned.Interface, annotations map[string]string) error {
+	return markannotation(tr, ps, annotations, ChainsAnnotation, Fail)
+}
+
+// AddRetries marks a TaskRun as retrying.
+func AddRetries(tr *v1beta1.TaskRun, ps versioned.Interface, annotations map[string]string) error {
```

Comment:
let's write this as:

```go
func AddRetries(tr *v1beta1.TaskRun, ps versioned.Interface, annotations map[string]string) error {
	retries := tr.ObjectMeta.Annotations[ChainsRetriesAnnotation]
	if retries == "" {
		return markannotation(tr, ps, annotations, ChainsRetriesAnnotation, "0")
	}
	v, err := strconv.Atoi(s)
	if err != nil {
		return errors.Wrapf(err, "invalid value for chain retries %s", s)
	}

	return markannotation(tr, ps, annotations, ChainsRetriesAnnotation, fmt.Sprintf("%d", v +1)
}
```

Reply:
we don't need to check `RetryAgain` because we already did before calling `AddRetries`